### ISSUE TITLE
type alias with typevar + --disallow-any-generic => unexpected error?

### ISSUE THREAD

**Issue by asottile** [python/mypy#5462]
This is a simplified example from a [larger project](https://github.com/asottile/all-repos/blob/46823f85e0eb78e15d57856afff6d873116f8604/all_repos/mapper.py).  I've resorted to spamming `Any` and avoiding the `TypeVar`s for now since type is longer than the code itself 😆 

```python
import contextlib
from typing import Callable
from typing import Generator
from typing import Iterable
from typing import TypeVar


T = TypeVar('T')
T2 = TypeVar('T2')
MC = Callable[[Callable[[T2], T], Iterable[T2]], Iterable[T]]


@contextlib.contextmanager
def f() -> Generator[Callable[[Callable[[T2], T], Iterable[T2]], Iterable[T]], None, None]:
    yield map


@contextlib.contextmanager
def f2() -> Generator[MC, None, None]:
    yield map
```

I expect `f` and `f2` to be identical, however:

```console
$ mypy t.py --disallow-any-generic
t.py:19: error: Missing type parameters for generic type
```

I can silence the error by doing:

```python
@contextlib.contextmanager
def f2() -> Generator[MC[T2, T], None, None]:
    yield map
```

or 

```python
@contextlib.contextmanager
def f2() -> Generator[MC[T, T2], None, None]:
    yield map
```

but it's unclear to me what that's doing and/or why it is different

**Comment by asottile** [comment]
Even with the trick with `MC[T, T2]`, the types are just wrong:

```python
import contextlib
from typing import Callable
from typing import Generator
from typing import Iterable
from typing import TypeVar


T = TypeVar('T')
T2 = TypeVar('T2')
MC = Callable[[Callable[[T2], T], Iterable[T2]], Iterable[T]]


@contextlib.contextmanager
def f() -> Generator[Callable[[Callable[[T2], T], Iterable[T2]], Iterable[T]], None, None]:
    yield map


@contextlib.contextmanager
def f2() -> Generator[MC[T2, T], None, None]:
    yield map


with f() as mapper:
    tuple(mapper(print, ('hello', 'there')))


with f2() as mapper:
    tuple(mapper(print, ('hello', 'there')))
```

```
$ mypy t.py
t.py:27: error: Incompatible types in assignment (expression has type "Callable[[Callable[[<nothing>], <nothing>], Iterable[<nothing>]], Iterable[<nothing>]]", variable has type "Callable[[Callable[[T2], T], Iterable[T2]], Iterable[T]]")
```

**Comment by ilevkivskyi** [comment]
This is not a bug, this is how generic aliases work. If you define:
```python
T1 = TypeVar('T1')
T2 = TypeVar('T2')
MC = Callable[[Callable[[T2], T1], Iterable[T2]], Iterable[T1]]
```
then _bare_ `MC` will be expanded with all type variables set to `Any` (same as any other generic):
```python
x: MC
reveal_type(x)  # Callable[[Callable[[Any], Any], Iterable[Any]], Iterable[Any]]
```
Instead, generic aliases could be used in a subscripted form, for example:
```python
y: MC[int, str]
reveal_type(y) # Callable[[Callable[[int], str], Iterable[str]], Iterable[int]]
```

Note also, that there is one subtlety with `Callable`, it binds the type variables to itself, but there is unfortunately a bug https://github.com/python/mypy/issues/3924.

I am closing this since as I explained the first post is as expected, while example in your second post is a duplicate of the mentioned issue.

**Comment by ilevkivskyi** [comment]
Btw, it looks like the generic part was just moved down (it was previously right after normal aliases). And this move was not done careful enough.

**Comment by asottile** [comment]
I can make a PR, how about something like:

```md
A type alias does not create a new type. It’s just a shorthand notation for another type – it’s equivalent to the target type unless it is a [generic type alias](...)
```

**Comment by ilevkivskyi** [comment]
OK

**Comment by asottile** [comment]
#5478 for that :)
